tests/scenario/test_tempo_api_implementation.py

```diff
@@ -0,0 +1,162 @@
+"""Tests that assert TempoCoordinatorCharm is wired up correctly to be a tempo-api provider."""
+
+import json
+from typing import Optional, Tuple
+from unittest.mock import PropertyMock, patch
+
+from ops.testing import Relation, State
+
+from tempo import Tempo
+
+RELATION_NAME = "tempo-api"
+INTERFACE_NAME = "tempo_metadata"
+
+TEMPO_COORDINATOR_URL = "http://needs-changing.com"
+INGRESS_URL = "http://www.ingress-url.com"
+
+
+def local_app_data_relation_state(
+    leader: bool, local_app_data: Optional[dict] = None
+) -> Tuple[Relation, State]:
+    """Return a testing State that has a single relation with the given local_app_data."""
+    if local_app_data is None:
+        local_app_data = {}
+    else:
+        # Scenario might edit this dict, and it could be used elsewhere
+        local_app_data = dict(local_app_data)
+
+    relation = Relation(RELATION_NAME, INTERFACE_NAME, local_app_data=local_app_data)
+    relations = [relation]
+
+    state = State(
+        relations=relations,
+        leader=leader,
+    )
+
+    return relation, state
+
+
+@patch(
+    "charm.TempoCoordinatorCharm._internal_url", PropertyMock(return_value=TEMPO_COORDINATOR_URL)
+)
+def test_provider_sender_sends_data_on_relation_joined(
+    context, s3, all_worker, nginx_container, nginx_prometheus_exporter_container
+):
+    """Tests that a charm using TempoApiProvider sends the correct data on a relation joined event."""
+    # Arrange
+    tempo_api = Relation(RELATION_NAME, INTERFACE_NAME)
+    relations = [
+        tempo_api,
+        s3,
+        all_worker,
+    ]
+
+    state = State(
+        relations=relations,
+        leader=True,
+        containers=[nginx_container, nginx_prometheus_exporter_container],
+    )
+
+    # Act
+    with context(context.on.relation_joined(tempo_api), state=state) as manager:
+        manager.run()
+        expected = {
+            "http": json.dumps(
+                {
+                    "direct_url": TEMPO_COORDINATOR_URL + f":{Tempo.server_ports['tempo_http']}/",
+                }
+            ),
+            "grpc": json.dumps(
+                {
+                    "direct_url": TEMPO_COORDINATOR_URL + f":{Tempo.server_ports['tempo_grpc']}/",
+                }
+            ),
+        }
+
+    # Assert
+    assert tempo_api.local_app_data == expected
```

Comment:
uhm I don't think this test is doing what you think it's doing.

`tempo_api` is immutable, and is the relation as you defined it before the event is processed.
What you need is:

```
state_out = context.run(context.on.relation_joined(tempo_api), state=state)
tempo_api_out = state_out.get_relation(tempo_api.id)
assert tempo_api.local_app_data == expected
```

Reply:
Turns out it did work, and [not on purpose](https://github.com/canonical/operator/issues/1605) :)  Refactored as suggested.